### ISSUE TITLE
Consistency of INSERT/SELECT with multiple clients

### ISSUE THREAD

**Issue by cpg314** [ClickHouse/ClickHouse#65020]
### Question

I am inserting one million rows into the following test table and then performing a `SELECT COUNT(*)`, using the [native protocol](https://clickhouse.com/docs/en/interfaces/tcp):
```sql
CREATE TABLE painter (`name` String)
ENGINE = ReplacingMergeTree
PRIMARY KEY name
ORDER BY name
```
```rust
client.insert(...).await;
let count = client.query(...).await; // 1 million
```
This works fine, and I am getting the right count:
```
11:36:13.622341 [ 39 ] {170d75a7-12ae-4138-96b6-bca45077d017} <Debug> executeQuery: (from 172.18.0.1:58576) INSERT INTO painter_local FORMAT native (stage: Complete)
11:36:14.191841 [ 39 ] {170d75a7-12ae-4138-96b6-bca45077d017} <Debug> executeQuery: Read 1500000 rows, 34.01 MiB in 0.569549 sec., 2633662.7752836016 rows/sec., 59.72 MiB/sec.
11:36:14.191880 [ 39 ] {170d75a7-12ae-4138-96b6-bca45077d017} <Debug> MemoryTracker: Peak memory usage (for query): 143.62 MiB.
11:36:14.191884 [ 39 ] {170d75a7-12ae-4138-96b6-bca45077d017} <Debug> TCPHandler: Processed in 0.569861306 sec.

11:36:14.192307 [ 39 ] {d57ab5aa-4f30-46e1-a7dc-31150a851c76} <Debug> executeQuery: (from 172.18.0.1:58576) SELECT COUNT(*) FROM painter_local  (stage: Complete)
11:36:14.193002 [ 682 ] {d57ab5aa-4f30-46e1-a7dc-31150a851c76} <Debug> MergingAggregatedTransform: Read 1 blocks of partially aggregated data, total 1 rows.
11:36:14.193042 [ 682 ] {d57ab5aa-4f30-46e1-a7dc-31150a851c76} <Debug> Aggregator: Converted aggregated data to blocks. 1 rows, 8.00 B in 8.11e-06 sec. (123304.562 rows/sec., 963.32 KiB/sec.)
11:36:14.193363 [ 39 ] {d57ab5aa-4f30-46e1-a7dc-31150a851c76} <Debug> executeQuery: Read 1 rows, 16.00 B in 0.001104 sec., 905.7971014492754 rows/sec., 14.15 KiB/sec.
11:36:14.193405 [ 39 ] {d57ab5aa-4f30-46e1-a7dc-31150a851c76} <Debug> TCPHandler: Processed in 0.001347153 sec.
```

However, when I do the INSERT then SELECT with two different clients (but _still sequentially_!) I get a count of 0. The data becomes visible to Client 2 only after 100-200ms.
```rust
client.1insert(...).await; // Wait until client1 finishes inserting
let count = client2.query(...).await; // 0
tokio::time::sleep(200ms).await;
let count = client2.query(...).await; // 1 million
```
```
# Client 1 starts inserting data
11:37:16.821954 [ 39 ] {08a866b8-fb68-4131-95cf-79d9705888d0} <Debug> executeQuery: (from 172.18.0.1:34902) INSERT INTO painter_local FORMAT native (stage: Complete)

# The INSERT call must have already returned, as Client 2 executes its COUNT!
11:37:17.158492 [ 40 ] {ac6342ee-5b27-4062-a56a-392511551e77} <Debug> executeQuery: (from 172.18.0.1:34918) SELECT COUNT(*) FROM painter_local  (stage: Complete)
11:37:17.202750 [ 722 ] {ac6342ee-5b27-4062-a56a-392511551e77} <Debug> MergingAggregatedTransform: Read 1 blocks of partially aggregated data, total 1 rows.
11:37:17.202782 [ 722 ] {ac6342ee-5b27-4062-a56a-392511551e77} <Debug> Aggregator: Converted aggregated data to blocks. 1 rows, 8.00 B in 4.995e-06 sec. (200200.200 rows/sec., 1.53 MiB/sec.)
11:37:17.203147 [ 40 ] {ac6342ee-5b27-4062-a56a-392511551e77} <Debug> executeQuery: Read 1 rows, 16.00 B in 0.044704 sec., 22.369362920544024 rows/sec., 357.91 B/sec.
11:37:17.203193 [ 40 ] {ac6342ee-5b27-4062-a56a-392511551e77} <Debug> TCPHandler: Processed in 0.044996675 sec.

# Only now the data from Client 1 finishes inserting. It makes sense that Client 2 did not see it.
11:37:17.392189 [ 39 ] {08a866b8-fb68-4131-95cf-79d9705888d0} <Debug> executeQuery: Read 1500000 rows, 34.01 MiB in 0.570289 sec., 2630245.366822786 rows/sec., 59.64 MiB/sec.
11:37:17.392222 [ 39 ] {08a866b8-fb68-4131-95cf-79d9705888d0} <Debug> MemoryTracker: Peak memory usage (for query): 143.62 MiB.
11:37:17.392225 [ 39 ] {08a866b8-fb68-4131-95cf-79d9705888d0} <Debug> TCPHandler: Processed in 0.570593744 sec.
```

Looking at the implementation of [the Rust client I use](https://github.com/Protryon/klickhouse/blob/master/klickhouse/src/client.rs#L350):
1. The client sends the insert query
2. The server returns a `TableColumns` packet and an empty `Data` packet
3. The client sends the data blocks.
4. The client sends a last data block, signalling the end of the transmission.

The clickhouse-go client has [the same logic](https://github.com/ClickHouse/clickhouse-go/blob/8ad6ec6b95d8b0c96d00115bc2d69ff13083f94b/conn_send_query.go#L26), likewise with the official [`clickhouse-client`](https://github.com/ClickHouse/ClickHouse/blob/873964b8829c6b10da6bb4669005c3a1cb6215ce/src/Client/ClientBase.cpp#L1727) 

But I do not see a packet back from the server to indicate that the data has been written.
In the single-client setup, this works because new writes on the connection seems to be blocked until the data insertion finishes; from the first logs above:
```
executeQuery: (from 172.18.0.1:58576) INSERT INTO painter_local FORMAT native (stage: Complete)
executeQuery: Read 1500000 rows, 34.01 MiB in 0.569549 sec., 2633662.7752836016 rows/sec., 59.72 MiB/sec.
TCPHandler: Processed in 0.569861306 sec.
```

If I make Client 1 wait for a `SELECT 1` to complete before Client 2 executes, then Client 2 sees all the rows.

In summary, am I right to believe that **the immediate visibility of INSERTs to subsequent queries is only true on the same client, and eventually for the others**?

**Comment by Algunenano** [comment]
```
client.1insert(...).await; // Wait until client1 finishes inserting
```

I think that this is not waiting until the end of the query and it's probably waiting only until the data is sent. The client is wrong and should wait and not close the connection, otherwise it won't know about the status of the query
